### PR TITLE
Add tests for compareAndSetValue and related unsafe APIs

### PR DESCRIPTION
**Pull request by ehrenjulzert**
part of testing for https://github.com/eclipse-openj9/openj9/issues/13614

Signed-off-by: Ehren Julien-Neitzert <ehren.julien-neitzert@ibm.com>

### CODE REVIEW

**Comment by ehrenjulzert** [comment]
@hangshao0 do you think what I'm doing with the `compareAndDoSomethingFuncs` data provider makes sense? It's the best way I could think of to test all 8 of the compareAndSetValue/compareAndExchangeValue functions.

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java [hunk 1]
```diff
@@ -80,6 +87,53 @@ static public void setUp() {
...
+		}
+
+		public String toString() {
+			return "CompareAndDoSomethingFunction(" + method.getName() + ")";
+		}
+	}
+
+	@DataProvider(name = "compareAndDoSomethingFuncs")
+	static public Object[][] compareAndDoSomethingFuncs() throws NoSuchMethodException {
+		return new Object[][] {
```
Comment (hangshao0) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:127]: You don't need all the wrappers here. I think `compareAndSetValue()` and `compareAndExchangeValue()` should be sufficient.

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java [hunk 2]
```diff
@@ -463,4 +514,156 @@ static public void testGetSizeOfZeroSizeVT() {
...
+			});
+		} else {
+			success = compareAndSwapValue.execute(vtPoint, vtPointOffsetX, ValueTypeInt.class, vtPoint.x, null);
+			assertTrue(success);
+			assertEquals(vtPoint.x, null);
+		}
+	}
+
+	@Test(dataProvider = "compareAndDoSomethingFuncs")
+	static public void testCompareAndSetObjectXSuccess(CompareAndDoSomethingFunction compareAndSwapValue) throws Throwable {
```
Comment (hangshao0) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:562]: The name of the function has "Object" in it. But the passed in parameter is "compareAndSwapValue". Seems confusing.

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java (hunk 1 above)
Comment (ehrenjulzert) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:127]: Okay, what about when I add tests for `getAndSetValue`, `getAndSetValueRelease`, and `getAndSetValueAcquire`? All 3 functions are almost identical so I could only write tests for `getAndSetValue` and ignore the others, or I could make another data provider for these 3 functions

Comment (hangshao0) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:127]: Test `getAndSetValue()` should be sufficient.

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java (hunk 2 above)
Comment (ehrenjulzert) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:562]: Okay, I can rename it to `testCompareAndSetPointXSuccess` to make it more clear

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java (hunk 1 above)
Comment (ehrenjulzert) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:127]: changed in 8add8f2

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java (hunk 2 above)
Comment (ehrenjulzert) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java]: updated in 8add8f2

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java [hunk 3]
```diff
@@ -80,6 +87,47 @@ static public void setUp() {
...
+			try {
+				returned = method.invoke(myUnsafe, obj, offset, clz, v1, v2);
+			} catch (InvocationTargetException exception) {
+				throw exception.getCause();
+			}
+			boolean result = false;
+			if (returned instanceof Boolean) {
+				result = (Boolean)returned;
+			} else {
+				result = returned == v1;
```
[the remark this answers is not in the document]
Comment (ehrenjulzert) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java]: fixed in 182972b

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java [hunk 4]
```diff
@@ -463,4 +508,242 @@ static public void testGetSizeOfZeroSizeVT() {
...
+			assertThrows(NullPointerException.class, () -> {
+				compareAndSwapValue.execute(null, vtPointOffsetX, ValueTypeInt.class, vtPoint.x, new ValueTypeInt(1));
+			});
+		}
+	}
+
+	@Test(dataProvider = "compareAndDoSomethingFuncs")
+	static public void testCompareAndSetNullClz(CompareAndDoSomethingFunction compareAndSwapValue) throws Throwable {
+		if (!isFlatteningEnabled) {
+			compareAndSwapValue.execute(vtPoint, vtPointOffsetX, null, vtPoint.x, new ValueTypeInt(1));
```
[the remark this answers is not in the document]
Comment (ehrenjulzert) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java]: added in 182972b

File: test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java [hunk 5]
```diff
@@ -463,4 +508,253 @@ static public void testGetSizeOfZeroSizeVT() {
 			assertEquals(size, 8);
 		}
 	}
+
+	@Test(dataProvider = "compareAndDoSomethingFuncs")
+	static public void testCompareAndSetNullObj(CompareAndDoSomethingFunction compareAndSwapValue) throws Throwable {
+		if (isFlatteningEnabled) {
```
Comment (hangshao0) [inline test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java:514]: I guess it segfaults in the `isFlatteningEnabled == false` case here ?

Comment (ehrenjulzert) [reply test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeUnsafeTests.java]: That's probably what I thought when I was writing the tests but it turns out it's just an NPE
I changed it to also test the unflattened case in cf98cd3

**Review by hangshao0 — APPROVED** [review]

**Comment by tajila** [comment]
Jenkins test sanity,extended zlinuxval jdknext

**Review by tajila — APPROVED** [review]

**Pull request merged**